test/helper_functions.jl

```diff
@@ -0,0 +1,33 @@
+@testset "helper functions" begin
+    # Hits fallback, since we can't update `Diagonal`s in place
```

Comment:
These blocks of code (e.g. `#2-6, 8:11, etc.`) should we re-written as nested `@test_sets`, this would increase readability, begin moving to tests being independent of each other, allow better documentation, and simplify running independent blocks.

Reply:
8-16 probably should be a single testsetset.

But in general yes.

Reply:
will do :)

Reply:
Looks good, ship it!